lib/explorer/series.ex

```diff
@@ -1253,8 +1267,14 @@ defmodule Explorer.Series do
           left :: Series.t(),
           right :: Series.t() | number() | Date.t() | NaiveDateTime.t() | boolean() | String.t()
         ) :: Series.t()
-  def equal(%Series{dtype: dtype} = left, %Series{dtype: dtype} = right),
-    do: apply_impl(left, :eq, [right])
+
+  def equal(%Series{dtype: _left_dtype} = left, %Series{dtype: _right_dtype} = right) do
+    case is_valid_length(left, right) do
+      {false, :zero_length} -> dtype_length_error("equal/2", left, right, :zero_length)
+      false -> dtype_length_error("equal/2", left, right)
+      true -> apply_impl(left, :eq, [right])
+    end
```

Comment:
Nice! I've already refactored it. But it will change the validation logic, hence valid patterns will return false so that invalid could return the reason.

I'm commiting it now.

Thanks for reviewing!